### ISSUE TITLE
Module export broken for UPDATE_OPS from AssignSub on a ResourceVariable

### ISSUE THREAD

**Issue by arnoegw** [tensorflow/hub#85]
## Scope

This issue is filed for the information of *creators* (not consumers) of modules that are trainable and contain UPDATE_OPS such as those used by exponential moving averages or batch normalization. 

This issue is already fixed in TensorFlow at head. Straightforward uses of `tf.layers.batch_normalization()` should be fine already in TensorFlow 1.9rc0.

Regarding concrete modules published on https://tfhub.dev/, a separate issue will be filed in a moment.

## Details

TensorFlow Hub modules created from code like

```python
def module_fn():
  ...
  v = tf.get_variable("v", ...)  # In variable_scope(..., use_resource=True) by default.
  update_op = tf.assign_sub(v, ...)
  tf.add_to_collection(tf.GraphKeys.UPDATE_OPS, update_op)
  ...
```

and used like

```python
m = hub.Module(..., tags={...}, trainable=True)
y = m(x)  # Brings in UPDATE_OPS that depend on x.
with tf.control_dependencies(tf.get_collection(tf.GraphKeys.UPDATE_OPS)):
  train_op = ...
with tf.Session() as session:
  session.run(train_op)
```

did not actually run `update_op` as expected if the module was created with a TensorFlow version before the recent patch https://github.com/tensorflow/tensorflow/commit/bb6532ae227b1f5ce72fbf3bd009f078698324de

The reason is that tf.assign_sub() on a ResourceVariable (the default variable types for Hub modules) does not return a plain Tensor, but a special object of the private type `_UnreadVariable`, which was improperly serialized into the UPDATE_OPS collection of a MetaGraphDef and hence a Hub module. (Non-resource variables are unaffected, but Hub Modules use ResourceVariables by default.)

The same issue applies to `tf.assign()` and `tf.assign_add()`.

`tf.assign_sub()` is used by the TensorFlow utilities to maintain exponential moving averages, including the mean and variance aggregated by batch normalization. Fortunately, `tf.layers.batch_normalization()` in TensorFlow 1.9rc0 ends up calling code in `tf.keras.layers.Layer` that extracts the underlying op from the result of `tf.assign_sub()` before putting it into UPDATE_OPS, which should avoid the issue. However, older or custom implementation of batch normalization may be at risk.

## How to check a module

If you have a model that might be affected, you can inspect its update ops as follows: Take an example use of it like in the code above that instantiates the trainable module and applies it to some example input. Then check that the op types in UPDATE_OPS are the kinds of variable updates that you expect.

```python
_VALID_UPDATE_OPS = {
    "AssignVariableOp", "AssignAddVariableOp", "AssignSubVariableOp",
    # Possibly more, depending on what you do,
    # but *not* "VarHandleOp" or "ReadVariableOp".
}

m = hub.Module(..., tags={...}, trainable=True)
y = m(x)  # Brings in UPDATE_OPS that depend on x.
for item in tf.get_collection(tf.GraphKeys.UPDATE_OPS):
  if isinstance(item, tf.Operation):
    update_op = item
  elif isinstance(item, tf.Tensor):
    update_op = item.op
  else:
    raise TypeError("Item %r of unexpected type" % item)

  if update_op.type not in _VALID_UPDATE_OPS:
    raise ValueError("Bad update op: %r" % update_op)
   
```
